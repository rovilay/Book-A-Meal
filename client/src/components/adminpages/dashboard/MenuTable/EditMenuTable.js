/* eslint jsx-a11y/label-has-for:0 */
/* eslint max-len:0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import sweetAlert from 'sweetalert';

class EditMenuTable extends Component {
  constructor(props) {
    super(props);

    this.getMealsNotInMenu = this.getMealsNotInMenu.bind(this);
    this.toggleMealCheckbox = this.toggleMealCheckbox.bind(this);
    this.updateMenu = this.updateMenu.bind(this);
  }

  /**
   * Checks meals for meals not currently in menu
   */
  getMealsNotInMenu() {
    const { content, meals } = this.props;
    const mealsInMenu = new Set(content.meals.map(meal => meal.id));
    return meals.filter(meal => !mealsInMenu.has(meal.id));
  }

  /**
   * check or unchecks meals
   * @param {*} meal meal to toggle
   */
  toggleMealCheckbox(meal) {
    return () => {
      const checkbox = document.getElementById(`${meal.id}-edit`);

      if (checkbox.checked === true) {
        this.props.addMealInEditMenu(meal.id);
      } else if (checkbox.checked === false) {
        this.props.deleteMealInEditMenu(meal.id);
      }
    };
  }

  /**
   * Updates menu
   * @param {*} event DOM event
   */
  /* istanbul ignore next */
  updateMenu(event) {
    event.preventDefault();
    const {
      hideModal,
      updateMenu,
      content,
      editMenuMeals
    } = this.props;
    const { menuId } = content;

    sweetAlert({
      text: 'Are you sure you want to update menu?',
      buttons: true,
      dangerMode: false,
    })
      .then((confirmed) => {
        if (confirmed) {
          const meals = [...editMenuMeals];
          updateMenu({ menuId, meals });
          hideModal();
        }
      })
      .catch(err => err);
  }

  render() {
    const {
      title,
      editMenuMeals,
    } = this.props;

    const mealsNotInMenu = this.getMealsNotInMenu();

    return (
      <div className="table-container">
        <h2 className="title">
          {title}
        </h2>

        <hr />
        <form
          onSubmit={this.updateMenu}
          className="editmenu-modal-form"
        >
          <div className="edit-menu-meals">
            <hr />
            {
              (mealsNotInMenu.length > 0)
                ?
                  <div className="checkbox-card">
                    {mealsNotInMenu.map(meal => (
                      <p key={meal.id} className="checkbox">
                        {meal.title}
                        <input
                          type="checkbox"
                          name="meal-check"
                          className="meal-check"
                          id={`${meal.id}-edit`}
                          onClick={this.toggleMealCheckbox(meal)}
                          value={meal.id}
                        />
                      </p>
                    ))}
                  </div>
                :
                  <div className="empty not-found">No meal to add!</div>
            }
          </div>

          {
            (mealsNotInMenu.length > 0)
            &&
            <div className="menu-updatebtn">
              <button
                type="submit"
                name="updateMenuBtn"
                id="menu-btn"
                className="btn-2"
                disabled={editMenuMeals.length <= 0}
              >
                Update Menu
              </button>
            </div>
          }
        </form>
      </div>
    );
  }
}

EditMenuTable.propTypes = {
  title: PropTypes.string.isRequired,
  addMealInEditMenu: PropTypes.func.isRequired,
  deleteMealInEditMenu: PropTypes.func.isRequired,
  meals: PropTypes.array.isRequired,
  hideModal: PropTypes.func.isRequired,
  editMenuMeals: PropTypes.array.isRequired,
  updateMenu: PropTypes.func.isRequired,
  content: PropTypes.object.isRequired
};

export default EditMenuTable;
