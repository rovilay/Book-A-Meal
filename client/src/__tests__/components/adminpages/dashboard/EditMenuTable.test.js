import React from 'react';
import { shallow } from 'enzyme';
import EditMenuTable from '../../../../components/adminpages/dashboard/MenuTable/EditMenuTable';
import { menuMeals, menus } from '../../../__mockData__/menuMock';

describe('EditMenuTable component test', () => {
  const setup = () => {
    const props = {
      title: "edit menu",
      addMealInEditMenu: () => jest.fn(),
      deleteMealInEditMenu: () => jest.fn(),
      meals: menuMeals,
      hideModal: () => jest.fn(),
      editMenuMeals: menuMeals,
      updateMenu: () => jest.fn(),
      content: {
        menuId: menus[0].id,
        meals: menuMeals.slice(0, 2)
      }
    };

    return shallow(<EditMenuTable {...props} />)
  };


  it('should render EditMenuTable correctly', (done) => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();

    done();
  });

  it('should call `toggleMealCheckbox`', (done) => {
    const wrapper = setup();

    document.body.innerHTML =
    `<div>
      <input type="checkbox" id="${menuMeals[0].id}-edit" checked="true" />
    </div>`;

    const toggleMealCheckboxSpy= jest.spyOn(wrapper.instance(), 'toggleMealCheckbox');
    wrapper.instance().toggleMealCheckbox(menuMeals[0])
    expect(toggleMealCheckboxSpy).toHaveBeenCalled();

    done();
  });

  it('should call `toggleMealCheckbox` on button click', (done) => {
    const wrapper = setup();

    document.body.innerHTML =
    `<div>
      <input type="checkbox" id="${menuMeals[3].id}-edit" checked="true" />
    </div>`;

    wrapper.find(`.meal-check`).last().simulate('click');
    const toggleMealCheckboxSpy= jest.spyOn(wrapper.instance(), 'toggleMealCheckbox');
    wrapper.instance().toggleMealCheckbox(menuMeals[3])
    expect(toggleMealCheckboxSpy).toHaveBeenCalled();

    done();
  });


  it('should call `updateMenu`', (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn()
    }

    const updateMenuSpy= jest.spyOn(wrapper.instance(), 'updateMenu');
    wrapper.instance().updateMenu(event)
    expect(updateMenuSpy).toHaveBeenCalled();

    done();
  });
});
