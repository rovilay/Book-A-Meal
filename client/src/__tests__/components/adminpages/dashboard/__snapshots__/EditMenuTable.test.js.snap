// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMenuTable component test should render EditMenuTable correctly 1`] = `
<div
  className="table-container"
>
  <h2
    className="title"
  >
    edit menu
  </h2>
  <hr />
  <form
    className="editmenu-modal-form"
    onSubmit={[Function]}
  >
    <div
      className="edit-menu-meals"
    >
      <hr />
      <div
        className="checkbox-card"
      >
        <p
          className="checkbox"
          key="3"
        >
          SPAGHETTI
          <input
            className="meal-check"
            id="3-edit"
            name="meal-check"
            onClick={[Function]}
            type="checkbox"
            value="3"
          />
        </p>
        <p
          className="checkbox"
          key="4"
        >
          YAM AND EGG
          <input
            className="meal-check"
            id="4-edit"
            name="meal-check"
            onClick={[Function]}
            type="checkbox"
            value="4"
          />
        </p>
      </div>
    </div>
    <div
      className="menu-updatebtn"
    >
      <button
        className="btn-2"
        disabled={false}
        id="menu-btn"
        name="updateMenuBtn"
        type="submit"
      >
        Update Menu
      </button>
    </div>
  </form>
</div>
`;
